jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -306,21 +299,45 @@ private void validateMemScope(ResourceScope memScope) throws IllegalStateExcepti
 	 * of the passed-in pointer argument to long.
 	 */
 	private final long memSegmtOfPtrToLongArg(MemorySegment argValue, LinkerOptions options) throws IllegalStateException {
-		UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue, options);
-		addMemArgScope(argValue.scope());
+		/*[IF JAVA_SPEC_VERSION >= 22]*/
+		HeapArgInfo info = heapArgInfo.get();
+		/*[ENDIF] JAVA_SPEC_VERSION >= 22 */
+
+		try {
+			UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue, options);
+			addMemArgScope(argValue.scope());
+		} catch (Throwable e) {
+			/*[IF JAVA_SPEC_VERSION >= 22]*/
+			/* Notify to reset the heap argument information for the current thread
+			 * due to the exception triggered here to guarantee the same handler with
+			 * different heap arguments passes through next time.
+			 */
+			if (info != null) {
+				info.index = 0;
+			}
+			/*[ENDIF] JAVA_SPEC_VERSION >= 22 */
+			throw e;
+		}
 
 		long address = argValue.address();
 		/*[IF JAVA_SPEC_VERSION >= 22]*/
+		if ((info == null) || (info.index == 0)) {
+			/* Remove the heap argument information last time when performing the downcall,
+			 * whether it ends up with a successful native invocation, an exception triggered
+			 * by validating the passed-in arguments or an exception from upcall captured in
+			 * native in which case a normal reset operation after the downcall won't handle.
+			 */
+			info = new HeapArgInfo(argLayoutArray.length);
+			heapArgInfo.set(info);
+		}
```

Comment:
With the addition of `HeapArgInfo.clear()` we don't need to check the index here:
```java
		if (info == null) {
			info = new HeapArgInfo(argLayoutArray.length);
			heapArgInfo.set(info);
		}
```

Reply:
Agreed and updated the code with the proposal above.